Avoid race condition during importer termination

alromeros opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:

This pull request sets the exit code when scratch space is required to 0 so, when deleting the importer pod after exiting for scratchSpaceNeeded, we don't need to deal with unwanted automatic restarts (RestartPolicyOnFailure). This conflicting behavior caused some issues during termination while the pod was already restarting after scratchSpaceNeeded.

Importer pod lifecycle for a regular DV before the fix:
```sh
# k get pod importer-cirros-dv-source -woyaml | grep exitCode
selecting docker as container runtime
        exitCode: 42
        exitCode: 42
        exitCode: 42
        exitCode: 42
        exitCode: 2
        exitCode: 42
        exitCode: 2
        exitCode: 42
        exitCode: 2
        exitCode: 0
        exitCode: 0
```
After the fix:
```sh
# k get pod importer-cirros-dv-source -woyaml | grep exitCode
selecting docker as container runtime
        exitCode: 0
        exitCode: 0
        exitCode: 0
        exitCode: 0
        exitCode: 0
        exitCode: 0
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # https://bugzilla.redhat.com/show_bug.cgi?id=2240963 (follow-up)

**Special notes for your reviewer**:

I think this is a better solution than https://github.com/kubevirt/containerized-data-importer/pull/3044 and https://github.com/kubevirt/containerized-data-importer/pull/3060. Let me know what you think.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Bugfix: Avoid race condition during importer termination
```

alromeros commented:
Third PR trying to accomplish this but I think this is the best solution yet. https://github.com/kubevirt/containerized-data-importer/pull/3060 led to nasty behavior when pod was failing with errors other than scratch space required and https://github.com/kubevirt/containerized-data-importer/pull/3044 0 grace period wasn't behaving well leaving containers stuck on the ceph lane.

alromeros commented:
/cc @akalenyu 
/cc @awels

akalenyu reviewed (commented):
I think this is a really good idea!

You are saying the way we have it today (non-zero) is plain wrong,
as up until now we were restarting indefinitely over the scratch exit code?

akalenyu commented:
/test pull-containerized-data-importer-e2e-nfs

akalenyu commented:
BTW - there is work to standardize the communication between importer & controllers at https://github.com/kubevirt/containerized-data-importer/pull/3103

alromeros commented:
> I think this is a really good idea!
> 
> You are saying the way we have it today (non-zero) is plain wrong, as up until now we were restarting indefinitely over the scratch exit code?

Yeah, when exiting with non-zero the pod keeps restarting indifinitely even when we just want to manually delete and recreate it. It isn't really an issue as it solves itself automatically but the deletion/creation is definitely racy.

akalenyu commented on pkg/controller/import-controller.go:
```diff
@@ -372,15 +372,14 @@ func (r *ImportReconciler) updatePvcFromPod(pvc *corev1.PersistentVolumeClaim, p
-		pod.Status.ContainerStatuses[0].LastTerminationState.Terminated.ExitCode > 0 {
-		log.Info("Pod termination code", "pod.Name", pod.Name, "ExitCode", pod.Status.ContainerStatuses[0].LastTerminationState.Terminated.ExitCode)
-		if pod.Status.ContainerStatuses[0].LastTerminationState.Terminated.ExitCode == common.ScratchSpaceNeededExitCode {
+		pod.Status.ContainerStatuses[0].State.Terminated != nil {
```
@0xFelix I am pretty sure there is no collision between this and https://github.com/kubevirt/containerized-data-importer/pull/3103/commits/8462345840cdd85410a4907fa9bd924326b17cd5, just making sure

akalenyu commented on pkg/controller/import-controller.go:
```diff
@@ -372,15 +372,14 @@ func (r *ImportReconciler) updatePvcFromPod(pvc *corev1.PersistentVolumeClaim, p
-		pod.Status.ContainerStatuses[0].LastTerminationState.Terminated != nil &&
-		pod.Status.ContainerStatuses[0].LastTerminationState.Terminated.ExitCode > 0 {
-		log.Info("Pod termination code", "pod.Name", pod.Name, "ExitCode", pod.Status.ContainerStatuses[0].LastTerminationState.Terminated.ExitCode)
-		if pod.Status.ContainerStatuses[0].LastTerminationState.Terminated.ExitCode == common.ScratchSpaceNeededExitCode {
```
So `LastTerminationState` is not being set because requiring scratch now results in a "successful" pod?

alromeros replied:
Right

akalenyu commented on pkg/controller/import-controller_test.go:
```diff
@@ -540,17 +540,10 @@ var _ = Describe("Update PVC from POD", func() {
 					State: v1.ContainerState{
 						Terminated: &corev1.ContainerStateTerminated{
-							ExitCode: 1,
-							Message:  "I went poof",
```
Why drop the message and reason? If I follow the code correctly they'll still get set

akalenyu reviewed (commented):
Just questions

/approve
/cc @mhenriks

alromeros commented:
/hold
Testing locally, works fine with rook-ceph-block but there is still an error code in the importer when using local storage class. Might be expected but I'll first make sure.

0xFelix reviewed (commented):
Actually, I think there is collision between this and https://github.com/kubevirt/containerized-data-importer/pull/3101. If the importer wants to signal that scratch space is required IMO it would be cleaner to do that with the proposed communication struct instead of matching the termination message string.

0xFelix commented:
During my work on #3101 and #3103 I also noticed that exit codes other than 0 do not work well in Kubernetes.

alromeros commented:
/unhold

@akalenyu Fixed the issue commented above: Returning 0 as normally clashed with cleanup functions that assume the imported file will be there during regular termination. Using `os.Exit(0)` instead avoids the issue.

akalenyu commented:
> /unhold
> 
> @akalenyu Fixed the issue commented above: Returning 0 as normally clashed with cleanup functions that assume the imported file will be there during regular termination. Using `os.Exit(0)` instead avoids the issue.

Ah, I see... how come this didn't fail any of the e2e tests?

alromeros commented:
> > /unhold
> > @akalenyu Fixed the issue commented above: Returning 0 as normally clashed with cleanup functions that assume the imported file will be there during regular termination. Using `os.Exit(0)` instead avoids the issue.
> 
> Ah, I see... how come this didn't fail any of the e2e tests?

Since the failing function is in `defer`, the scratch space required termination message was still written. After a couple of restarts, the pod with scratch space was created and the import succeeded anyway.

alromeros commented:
/retest-required

0xFelix commented:
Is there a chance for this to be part of #3101? IMO both this PR and #3101 try to solve very similar issues.

alromeros commented:
> Is there a chance for this to be part of #3101? IMO both this PR and #3101 try to solve very similar issues.

Hey @0xFelix, yeah it makes sense to eventually integrate this into #3101, probably using a specific field (`ScratchSpaceRequired` boolean maybe?) to avoid parsing the msg in the controller. That said, I would prefer to prioritize merging this first since this is part of a bugfix we plan to backport to v1.57, and it'd be safer to avoid backporting a larger feature. Thanks!

0xFelix commented:
Alright, makes sense if you want to backport this.

alromeros commented:
Added a new commit (https://github.com/kubevirt/containerized-data-importer/pull/3116/commits/3db5b724e4dd9612191a002a7d2f1930c50ca163) to address a flake caused by the new behavior:

The test relied on the assumption that deleting the img from the http server would always cause the DV to restart at least once. However, with the new faster recovery time in the importer, a race condition happened where the file was deleted with the download already started, which just caused the polling to keep retrying without failing. Since we recreate the file fast enough, there was no time for the DV to error,  so no restart was needed.

This test could also rely on false positives since the importer pod failing for scratch space always caused the DV to restart.

To fix this I deleted the part of the test that checks for dv restarts to be >= 1, and to make sure that everything is working as expected I added a md5sum check at the end.

akalenyu commented:
> Added a new commit ([3db5b72](https://github.com/kubevirt/containerized-data-importer/commit/3db5b724e4dd9612191a002a7d2f1930c50ca163)) to address a flake caused by the new behavior:
> 
> The test relied on the assumption that deleting the img from the http server would always cause the DV to restart at least once. However, with the new faster recovery time in the importer, a race condition happened where the file was deleted with the download already started, which just caused the polling to keep retrying without failing. Since we recreate the file fast enough, there was no time for the DV to error, so no restart was needed.
> 
> This test could also rely on false positives since the importer pod failing for scratch space always caused the DV to restart.
> 
> To fix this I deleted the part of the test that checks for dv restarts to be >= 1, and to make sure that everything is working as expected I added a md5sum check at the end.

Makes sense
/test pull-containerized-data-importer-e2e-nfs

akalenyu commented:
/test pull-containerized-data-importer-fossa
weird one, fossa yells about `github.com/gorhill/cronexpr` which has been around for ages

akalenyu commented:
/test pull-containerized-data-importer-e2e-nfs

akalenyu commented:
/test pull-containerized-data-importer-e2e-nfs

akalenyu commented:
/test pull-containerized-data-importer-fossa

mhenriks commented:
/lgtm

alromeros commented:
There's been some conversations around the possibility of not backporting this. I'll leave it like this until we make a decision.

akalenyu commented:
> There's been some conversations around the possibility of not backporting this. I'll leave it like this until we make a decision.

Yeah not a fan of backporting this, it's a subtle change, so unless someone is desperately asking for a backport I wouldn't do it
